ospp/OSPP-2025.md

```diff
@@ -78,3 +78,88 @@ Libraries & Tools:
+
+## Project Description
+
+The virtio-sound device emulation has been evolving to provide better virtualization support for audio functionalities.
+The goal of this project is to extend the capabilities of the virtio-sound device within the [vhost-device-sound](https://github.com/rust-vmm/vhost-device/tree/main/vhost-device-sound)
+crate by adding audio control features such as volume adjustment, muting/unmuting, and mode switching.
+
+Recent [patches](https://lore.kernel.org/all/20240423150951.41600-39-pbonzini@redhat.com/) have introduced support for audio
+controls in the Linux headers used by QEMU, and the [VirtIO sound 1.3 specification](https://docs.oasis-open.org/virtio/virtio/v1.3/csd01/virtio-v1.3-csd01.html#x1-56700014)
+has now incorporated these audio controls feature.
+
+This project will focus on implementing these new features in the Rust-based vhost-device-sound crate. This new feature enhancement will
+allow a user of the guest OS to dynamically control audio settings, improving the flexibility and usability of virtualized audio devices.
+
+The audio controls support should be implemented following the [VirtIO sound 1.3 specification](https://docs.oasis-open.org/virtio/virtio/v1.3/csd01/virtio-v1.3-csd01.html#x1-56700014).
+
+## Project Notes
+
+- This project should extend vhost-device-sound to support audio control operations in compliance with VirtIO Sound 1.3 specification.
+- Add Feature Negotiation support for [VIRTIO_SND_F_CTLS](https://gitlab.com/qemu-project/qemu/-/blob/master/include/standard-headers/linux/virtio_snd.h#L15)
+  to both vhost-device-sound and QEMU to ensure that control messages are only enabled if `VIRTIO_SND_F_CTLS` is successfully negotiated between the guest and host.
+- Extend the vhost-device-sound to Handle the new audio control messages from the guest. 
+- Requires modifications to QEMU’s vhost-user-sound frontend to handle feature negotiation and control messages.
+- The implementation should support:
+    - Volume control (per-channel adjustments).
+    - Muting/unmuting audio streams.
+    - Switching audio output modes (e.g., speakers/headphones).
+- Extend the [PipeWire Backend](https://github.com/rust-vmm/vhost-device/blob/main/vhost-device-sound/src/audio_backends/pipewire.rs)
+  to Implement audio control features using the [pipewire-rs crate](https://gitlab.freedesktop.org/pipewire/pipewire-rs).
+- Extend the [Alsa Backend](https://github.com/rust-vmm/vhost-device/blob/main/vhost-device-sound/src/audio_backends/alsa.rs)
+  to Implement audio control features using the [alsa rust bindings](https://github.com/diwic/alsa-rs.git).
+- The implementation will be tested using QEMU as the vhost-user frontend.
+- Automated tests will be written using cargo test.
+- A possible stretch goal is contributing to rust-vmm/vhost-device by fixing existing issues.
+
+## Project Difficulty
+
+Advanced (some understanding of QEMU pci device would be good to know)
```

Comment:
Here IIUC QEMU need to be touched, so it's fine to mention, but it's not clear what "pci device" means. Are you referring to generic PCI support in QEMU, or the vhost-user sound frontend?

Reply:
Just the hos-user sound frontend. ok I will make that clear thanks!